clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -423,34 +426,31 @@ public synchronized TransactionalRequestResult sendOffsetsToTransaction(final Ma
         return handler.result;
     }
 
-    public synchronized void maybeAddPartitionToTransaction(TopicPartition topicPartition) {
-        if (isPartitionAdded(topicPartition) || isPartitionPendingAdd(topicPartition))
-            return;
+    public synchronized void maybeAddPartition(TopicPartition topicPartition) {
+        maybeFailWithError();
+        throwIfPendingState("send");
 
-        log.debug("Begin adding new partition {} to transaction", topicPartition);
-        topicPartitionBookkeeper.addPartition(topicPartition);
-        newPartitionsInTransaction.add(topicPartition);
+        if (isTransactional()) {
+            if (!hasProducerId()) {
+                throw new KafkaException("Cannot add partition " + topicPartition +
+                    "to transaction before completing a call to initTransactions");
+            } else if (currentState != State.IN_TRANSACTION) {
+                throw new KafkaException("Cannot add partition " + topicPartition +
```

Comment:
Why are we changing to using `KafkaException` here? I understand that you use `KafkaException` consistently in all the cases now but why?

It might be worth checking the javadoc of the methods in `KafkaProducer` if we stick to `KafkaException`.

Reply:
Consistency was the only reason. It looked weird in the test case.